### ISSUE TITLE
Metadata Block for MIME Type

### ISSUE THREAD

**Issue by sjehuda** [greasemonkey/greasemonkey#3137]
There are userscripts that intended for [XML documents](https://github.com/greasemonkey/greasemonkey/issues/3135) or HTML documents, or even [Image documents](https://github.com/Owyn/HandyImage) , some exclusively and some for multiple types.

Currently,  I use `Document.contentType` (e.g. `document.contentType.match('xml')` to determine whether or not script should start.

Metadata block API can make this task easier for scriptwriters and make a friendlier user experience.

**Comment by sjehuda** [comment]
There are userscripts that intended for [XML documents](https://github.com/greasemonkey/greasemonkey/issues/3135) or HTML documents, or even [Image documents](https://github.com/Owyn/HandyImage) , some exclusively and some for multiple types.

Currently,  I use `Document.contentType` (e.g. `document.contentType.match('xml'))` to determine whether or not script should start.

However, a better way to find out the filetype, is by `XMLHttpRequest`.
```
xmlHTTPRequest = new XMLHttpRequest();
xmlHTTPRequest.open('GET', document.documentURI, false);
xmlHTTPRequest.send();
xmlHTTPRequest.getResponseHeader('Content-Type');
```
But, currently, it's not allowed (deprecated) to make such request on the main thread unless it's inside a so-called Web Worker, which must be inside a separate JS file.

Surely, we can call a file as string and manipulate it, but it's not a proper solution, IMHO.

A built-in functionality (extension level) that is able to store filetype upon request is vital, of course only when that metadata block is used.

Metadata block API can make this task easier for scriptwriters and make a friendlier user experience.

**Comment by sjehuda** [comment]
In my case, it will also help to be avoided from being blocked by server due to "surge of (http) requests".  And considering the nature of Greasemonkey, this feature is a must!

**Comment by arantius** [comment]
Your request doesn't make total sense.  You shouldn't be making additional requests to decide whether to run your code, it's better to use the feature you pointed out (`document.contentType`) to decide if you want to process an already loaded document.

That said: the general policy of Greasemonkey is *not* to implement custom features when the user script is already capable of doing so on its own, which it looks like you've documented one or two ways already above.

**Comment by sjehuda** [comment]
I agree, but when we take into consideration the fact that that Greasemonkey was made to work with 3rd party websites, then it makes a conplete sense.

Take this Web Feed as an example: https://otter-browser.org/feed/

When we run a check using "XMLHttpRequest"
```
xmlHTTPRequest = new XMLHttpRequest();
xmlHTTPRequest.open('GET', document.documentURI, false);
xmlHTTPRequest.send();
xmlHTTPRequest.getResponseHeader('Content-Type');
```
**"application/atom+xml; charset=utf-8"**

When we run a check using "document" we might get
```
document.contentType
```
**"text/html"**

Should I call the Otter Browser website sysadmin to correct his server configurations?
Sure, I can do that, but this is not only on his websites, as there are plenty of websites with similar case.

As I wrote above:
1. Greasemonkey essentially intends to provides the users anything that 3rd party websites don't.
2. I want to avoid as much as possible from sending extra HTTP requests, due to the risk of being blocked and also privacy.

**Comment by sjehuda** [comment]
I agree, but when we take into consideration the fact that Greasemonkey was made to work with 3rd party websites, then it makes a complete sense.

Take this Web Feed as an example: https://otter-browser.org/feed/

When we run a check using "XMLHttpRequest"
```
xmlHTTPRequest = new XMLHttpRequest();
xmlHTTPRequest.open('GET', document.documentURI, false);
xmlHTTPRequest.send();
xmlHTTPRequest.getResponseHeader('Content-Type');
```
**"application/atom+xml; charset=utf-8"**

When we run a check using "document" we might get
```
document.contentType
```
**"text/html"**

Should I call the Otter Browser website sysadmin to correct his server configurations?
Sure, I can do that, but this is not only on his websites, as there are plenty of websites with similar case.
Will he be willing to make that change?

As I wrote above:
1. Greasemonkey essentially intends to provides the users anything that 3rd party websites don't.
2. I want to avoid as much as possible from sending extra HTTP requests, due to the risk of being blocked and also privacy.

_P.S. Before I have been migrating to Falkon web browser, I was using your Resurrect Pages extension  - IMHO, it's simple and briliant!_

**Comment by sjehuda** [comment]
I agree, but when we take into consideration the fact that Greasemonkey was made to work with 3rd party websites, then it makes a complete sense.

Take this Web Feed as an example: https://otter-browser.org/feed/

When we run a check using "XMLHttpRequest"
```
xmlHTTPRequest = new XMLHttpRequest();
xmlHTTPRequest.open('GET', document.documentURI, false);
xmlHTTPRequest.send();
xmlHTTPRequest.getResponseHeader('Content-Type');
```
**"application/atom+xml; charset=utf-8"**

When we run a check using "document" we might get
```
document.contentType
```
**"text/html"**

Should I call the Otter Browser website sysadmin to correct his server configurations?
Sure, I can do that, but this is not only on his websites, as there are plenty of websites with similar case.
Will he be willing to make that change?

As I wrote above:
1. Greasemonkey essentially intends to provides the users anything that 3rd party websites don't.
2. I want to avoid as much as possible from sending extra HTTP requests, due to the risk of being blocked and also privacy.

_P.S. Before I have been migrating to Falkon web browser, I was using your Resurrect Pages extension  - IMHO, it's simple and briliant!_

**Comment by sjehuda** [comment]
> Should I call the Otter Browser website sysadmin to correct his server configurations?
> Sure, I can do that, but this is not only on his websites, as there are plenty of websites with similar case.
> Will he be willing to make that change?

I ask you to observe the above quote.
The results between `document.contentType` and `xmlHTTPRequest.getResponseHeader('Content-Type');` may vary.

1. Sending an extra request is, in fact, an additional fingerprint which may disclose users identity.
2. Sending an extra request might as well damage performance (e.g. HTML or XML file is above 1MB and connection is to slow to load this in a reasonable amount of time).
3. The nature of Greasemonkey (and extensions in general), as far as I'm concerned over all these years, is to provide the UX that will not be provided by 3rd party websites, be it by will or becuse of constrains.

Please, think this think through, again.

Thank you.

**Comment by sjehuda** [comment]
Problem solved.

I need to change `@run-at` to `document-start` and call twice to XMLHTTPRequest, one with HEAD and `XMLHttpRequest.readyState === 2` (HEADERS_RECEIVED) and if page is a Feed, then once again with GET and `XMLHttpRequest.readyState === 3` (LOADING) or `4` (DONE) when MIME-Type doesn't say ATOM or RSS but we find out that it is.
